### DISCUSSION TITLE
Expose output format during cell execution

### DISCUSSION THREAD

**Discussion by kadykov** [quarto-dev/quarto-cli#9303] [Q&A]
## Background

I have a document that I render in `html`, `pdf` and `docx` formats.
I use Python code cells like this to render figures with `hvplot` and `xarray`:

````qmd
```{python}
#| fig-cap: "Tiny example"
import os

import xarray as xr
import hvplot
import hvplot.xarray

ds = xr.tutorial.open_dataset("tiny")

ds.tiny.hvplot()
```
````

## The Problem

Since `pdf` and `docx` do not support rendering `bokeh` figures,
I have to switch to `matplotlib` backend for these output formats.
The solution that I use right now is:

````qmd
::: {.content-visible when-format="html"}
```{python}
#| fig-cap: "Tiny example"
# The same code
...
ds.tiny.hvplot()
```
:::

::: {.content-visible when-format="pdf"}
```{python}
#| fig-cap: "Tiny example"
# The same code
...
# Use `matplotlib` backend
hvplot.extension("matplotlib")
ds.tiny.hvplot()
```
:::

::: {.content-visible when-format="docx"}
```{python}
#| fig-cap: "Tiny example"
# The same code
...
# Use `matplotlib` backend
hvplot.extension("matplotlib")
ds.tiny.hvplot()
```
:::
````

As you see, I have to repeat the same code block three times with very minor changes.
Please correct me if there is a better solution to this problem.

## Expected solution

If we could have access to the output format during script execution,
this code could be much simpler:

````qmd
```{python}
#| fig-cap: "Tiny example"
import os

import xarray as xr
import hvplot
import hvplot.xarray

if os.environ.get("QUARTO_OUTPUT_FORMAT") == "html":
    # Use Matplotlib backend for everything except html output
    hvplot.extension("matplotlib")

ds = xr.tutorial.open_dataset("tiny")

ds.tiny.hvplot()
```
````

> There is already `QUARTO_FIG_FORMAT` environment variable, but it cannot be used for this purpose, as for `html` and `docx` formats it has the same value = `png`. For `pdf` it has the value `pdf`.

quarto: 1.4.553
python: 3.11.9
os: Win10

**Comment by mcanouil** [comment]
Could you properly format your post using code blocks for code and terminal outputs? Thanks.
If your code contains code blocks, you need to enclose it using more backticks, *i.e.*, usually four ` ```` `.

Currently none of your code is usable.


---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

If you have multiple files (and if it is **absolutely required** to have multiple files), please share as a Git repository.

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image](https://placehold.co/600x400.png)

The end.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image](https://placehold.co/600x400.png)

The end.
````
`````

</td></tr>
</table>

Additionally and if not already given, please share the output of `quarto check` within a code blocks (*i.e.*, using three backticks ` ```txt `), see <https://quarto.org/bug-reports.html#check>.

**Comment by mcanouil** [comment]
Jupyter does not have the concept of reusable code cell.
What you can do is to externalise the code cell and use the include shortcode. This way you repeat the include but not the code.

**Comment by cscheid** [comment]
We will implement something like this in the future. No concrete timeline to give you yet, but it's something we know we have to do.
